src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+    ///            and "x = Computation();" is the first statement in the method body, then it's initial value
+    ///            is never used. Such a parameter should be removed and 'x' should be converted into a local.
+    ///        We provide additional information in the diagnostic message to clarify the above two categories
+    ///        and also detect and mention about potential breaking change if the containing method is a public API.
+    ///        Currently, we do not provide any code fix for removing unused parameters as it needs fixing the
+    ///        call sites and any automated fix can lead to subtle overload resolution differences,
+    ///        though this may change in future.
+    /// </summary>
+    internal abstract class AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private const string UnusedValuePreferenceKey = nameof(UnusedValuePreferenceKey);
+        private const string IsUnusedLocalAssignmentKey = nameof(IsUnusedLocalAssignmentKey);
+        private const string IsRemovableAssignmentKey = nameof(IsRemovableAssignmentKey);
+
+        // "Expression value is never used"
+        private static readonly DiagnosticDescriptor s_expressionValueIsUnusedRule = CreateDescriptorWithId(
+            IDEDiagnosticIds.ExpressionValueIsUnusedDiagnosticId,
+            new LocalizableResourceString(nameof(FeaturesResources.Expression_value_is_never_used), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            new LocalizableResourceString(nameof(FeaturesResources.Expression_value_is_never_used), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: false);
+
+        // "Value assigned to '{0}' is never used"
+        private static readonly DiagnosticDescriptor s_valueAssignedIsUnusedRule = CreateDescriptorWithId(
+            IDEDiagnosticIds.ValueAssignedIsUnusedDiagnosticId,
+            new LocalizableResourceString(nameof(FeaturesResources.Value_assigned_to_symbol_is_never_used), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            new LocalizableResourceString(nameof(FeaturesResources.Value_assigned_to_0_is_never_used), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: true);
+
+        // "Remove unused parameter '{0}'{1}"
+        private static readonly DiagnosticDescriptor s_unusedParameterRule = CreateDescriptorWithId(
+            IDEDiagnosticIds.UnusedParameterDiagnosticId,
+            new LocalizableResourceString(nameof(FeaturesResources.Remove_unused_parameter), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            new LocalizableResourceString(nameof(FeaturesResources.Remove_unused_parameter_0_1), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+            isUnneccessary: true);
+
+        private static readonly PropertiesMap s_propertiesMap = CreatePropertiesMap();
```

Comment:
i see now what this is used for.  but it really needs a comment explaining things.